jni/src/linear-types/Foreign/JNI/Linear.hs

```diff
@@ -0,0 +1,472 @@
+-- | Low-level bindings to the Java Native Interface (JNI).
+--
+-- This module provides an interface with linear types to the functions
+-- of JNI. All local references to objects are tracked so the compiler
+-- ensures that they are eventually deleted and reasonably fast.
+--
+-- Class references and global references aren't tracked because their
```

Comment:
Class references are not global. We can state that explicitly, indeed.  The call methods could return class references, and it wouldn't be trivial to discriminate them to make them global.

Reply:
On second thought, maybe we could have our class producing functions `findClass` and `getObjectClass` produce global references. If the call methods return a class, the caller will know it is not global because it has a different type.